pulsar-io/kafka/src/main/java/org/apache/pulsar/io/kafka/AvroSchemaCache.java

```
@@ -0,0 +1,104 @@
+import io.confluent.kafka.schemaregistry.client.rest.exceptions.RestClientException;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.pulsar.client.api.Schema;
+import org.apache.pulsar.common.schema.SchemaInfo;
+import org.apache.pulsar.common.schema.SchemaType;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.Collections;
+import java.util.concurrent.ExecutionException;
+
+@Slf4j
+final class AvroSchemaCache {
+    private final LoadingCache<Integer, Schema<ByteBuffer>> cache = CacheBuilder
+            .newBuilder()
+            .maximumSize(100)
+            .build(new CacheLoader<Integer, Schema<ByteBuffer>>() {
+                @Override
+                public Schema<ByteBuffer> load(Integer schemaId) throws Exception {
+                    return fetchSchema(schemaId);
+                }
+            });
+
+    private final SchemaRegistryClient schemaRegistryClient;
+
+    public AvroSchemaCache(SchemaRegistryClient schemaRegistryClient) {
+        this.schemaRegistryClient = schemaRegistryClient;
+    }
+
+    public Schema<ByteBuffer> get(int schemaId) {
+        try {
+            return cache.get(schemaId);
+        } catch (ExecutionException err) {
+            throw new RuntimeException(err.getCause());
+        }
+    }
+
+    private Schema<ByteBuffer> fetchSchema(int schemaId) {
+        try {
+            org.apache.avro.Schema schema = schemaRegistryClient.getById(schemaId);
+            String definition = schema.toString(false);
+            log.info("Schema {} definition {}", schemaId, definition);
+            SchemaInfo schemaInfo = SchemaInfo.builder()
+                    .type(SchemaType.AVRO)
+                    .name(schema.getName())
+                    .properties(Collections.emptyMap())
+                    .schema(definition.getBytes(StandardCharsets.UTF_8)
+                    ).build();
+            return new Schema<ByteBuffer>() {
```

I think you need to override the decode method and throw an UnsupportedException. Otherwise, if the decode method is used, it will cause StackOverflowException.

---

done